js/viz.js

```diff
@@ -110,35 +110,105 @@ const viz = {
     for (const node of this.nodes) {
       this.context.beginPath();
       this.context.moveTo(node.x, node.y);
+
+      if (node.shadow) {
+        this.drawShadow(node.x, node.y);
+      }
+
       if (node.firstParty) {
         this.drawFirstParty(node);
       } else {
         this.drawThirdParty(node);
       }
+
+      if (node.favicon) {
+        this.drawFavicon(node.favicon, node.x, node.y);
+      }
+
       this.context.fillStyle = 'white';
       this.context.closePath();
       this.context.fill();
     }
   },
 
+  getSquare() {
+    const side = Math.sqrt(this.circleRadius * this.circleRadius * 2);
+    const offset = side * 0.5;
+
+    return {
+      side,
+      offset
+    };
+  },
+
+  drawFavicon(favicon, x, y) {
+    const { side, offset } = this.getSquare();
+    const img = new Image();
+    img.src = favicon;
+    img.onload = () => {
+      this.context.drawImage(img,
+        this.transform.applyX(x) - offset,
+        this.transform.applyY(y) - offset,
+        side,
+        side);
+    };
+  },
+
+  drawShadow(x, y) {
+    this.context.beginPath();
+    this.context.lineWidth = 6;
+    this.context.shadowColor = 'white';
+    this.context.strokeStyle = 'rgba(0,0,0,1)';
+    this.context.shadowBlur = 15;
+    this.context.shadowOffsetX = 0;
+    this.context.shadowOffsetY = 0;
+    this.context.arc(x, y, this.circleRadius + 5, 0, 2 * Math.PI);
+    this.context.stroke();
+    this.context.closePath();
+  },
+
   drawFirstParty(node) {
     this.context.arc(node.x, node.y, this.circleRadius, 0, 2 * Math.PI);
   },
 
   drawThirdParty(node) {
-    // draw third party as equilateral triangle
     const deltaY = this.circleRadius / 2;
     const deltaX = deltaY * Math.sqrt(3);
+
     this.context.moveTo(node.x - deltaX, node.y + deltaY);
     this.context.lineTo(node.x, node.y - this.circleRadius);
     this.context.lineTo(node.x + deltaX, node.y + deltaY);
   },
 
+  getTooltipPosition(x, y) {
+    const { right: canvasRight } = this.canvas.getBoundingClientRect();
+    const {
+      height: tooltipHeight,
+      width: tooltipWidth
+    } = this.tooltip.getBoundingClientRect();
+    const top = y - tooltipHeight - this.circleRadius - 20;
+
+    let left;
+    if ((x + tooltipWidth) >= canvasRight
```

Comment:
One of them is redundant unless you are running this code multiple times perhaps?

When `x + (tooltipWidth/2)) >= canvasRight` is true `(x + tooltipWidth) >= canvasRight` will always be true.